### PR TITLE
ARROW-4407: [C++] Cache compiler for CMake external projects

### PR DESCRIPTION
**Pull request by fsaintjacques**
Ensure that external projects use the same compiler that was first
defined when invoking the `cmake` call.

### CODE REVIEW

**Review by wesm — COMMENTED** [review]
Interesting. I guess we haven't run into this generally because of using the `$CC` and `$CXX` environment variables, is that right?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -500,6 +500,8 @@ if("${DOUBLE_CONVERSION_HOME}" STREQUAL "")
 
   set(DOUBLE_CONVERSION_CMAKE_ARGS
         "-DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}"
+        "-DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}"
+        "-DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}"
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: While I'm at it, should I move `CMAKE_BUILD_TYPE` in that common flags too?

**Comment by fsaintjacques** [comment]
I've cleaned up a lot of accumulated cruft. One issue is that I haven't tested the Apple/Msvc branches.

**Comment by fsaintjacques** [comment]
Ping, should be ready after a push-force.

**Review by wesm — APPROVED** [review]
+1. I'm taking this for a spin on Windows before merging since that's the most likely configuration to fail

**Comment by wesm** [comment]
looks good

**Pull request closed without merging**